src/error.rs

```diff
@@ -1,50 +1,76 @@
-use std::borrow::Cow;
-use std::num::ParseIntError;
-use std::string::ParseError;
 use std::{error, fmt, io};
 
-#[derive(Clone, Debug)]
-pub struct Error {
-    message: Cow<'static, str>,
+/// A program error.
+///
+/// Such errors are highly likely to halt the program.
+#[derive(Clone)]
+pub struct Error(Box<Information>);
+
+/// Information about an error.
+#[derive(Clone)]
+struct Information {
+    /// The primary error message.
+    ///
+    /// This is the innermost error to occur.
+    primary: Box<str>,
+
+    /// Layers of context to the error.
+    ///
+    /// Ordered from innermost to outermost.
+    context: Vec<Box<str>>,
 }
 
-impl From<&'static str> for Error {
-    fn from(message: &'static str) -> Self {
-        Self {
-            message: Cow::Borrowed(message),
-        }
+impl Error {
+    /// Construct a new error from a string.
+    pub fn new(error: &str) -> Self {
+        Self(Box::new(Information {
+            primary: error.into(),
+            context: Vec::new(),
+        }))
     }
-}
 
-impl From<String> for Error {
-    fn from(message: String) -> Self {
-        Self {
-            message: Cow::Owned(message),
-        }
+    /// Add context to this error.
+    pub fn context(mut self, context: &str) -> Self {
+        self.0.context.push(context.into());
+        self
     }
 }
 
-impl From<ParseError> for Error {
-    fn from(_err: ParseError) -> Self {
-        Self::from("message parse error")
+impl From<&str> for Error {
+    fn from(error: &str) -> Self {
+        Self::new(error)
     }
 }
 
-impl From<ParseIntError> for Error {
-    fn from(_err: ParseIntError) -> Self {
-        Self::from("message parse error")
+impl From<String> for Error {
+    fn from(error: String) -> Self {
+        Self::new(&error)
     }
 }
 
 impl From<io::Error> for Error {
-    fn from(err: io::Error) -> Self {
-        Self::from(err.to_string())
+    fn from(error: io::Error) -> Self {
+        Self::new(&error.to_string())
     }
 }
 
 impl fmt::Display for Error {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        fmt::Display::fmt(&self.message, f)
+        f.write_str("\x1E[31mERROR:\x1E[0m ")?;
```

Comment:
The coreutils always print their own executable name (i.e. `argv[0]`), which is very nice for scripts with multiple things happening. Might be nice to do that as well.